File: problems/0006_zigzag-conversion/Solution.scala
object Solution {
    def convert(s: String, numRows: Int): String = {
        if (s.size <= 2 || numRows == 1)
            return s
        val colsPerPad = numRows - 1
        val charsPerPad = colsPerPad - 1 + numRows
        val lastRow = colsPerPad
        val buf = new StringBuilder(s.size)
        for (r <- 0 until numRows)
            for (i <- r until s.size by charsPerPad) {
                buf += s(i)
                val j = i + charsPerPad - 2 * r
                if (r != 0 && r != lastRow && j < s.size)
                    buf += s(j)
            }
        buf.toString
    }
}

File: problems/0006_zigzag-conversion/Solution.cs
using System.Text;

public class Solution {
    public string Convert(string s, int numRows) {
        if (s.Length <= 2 || numRows == 1)
            return s;
        int colsPerPad = numRows - 1,
            charsPerPad = colsPerPad - 1 + numRows,
            lastRow = colsPerPad;
        var buf = new StringBuilder(s.Length);
        for (var r = 0; r < numRows; ++r)
            for (var i = r; i < s.Length; i += charsPerPad) {
                buf.Append(s[i]);
                var j = i + charsPerPad - 2 * r;
                if (r != 0 && r != lastRow && j < s.Length)
                    buf.Append(s[j]);
            }
        return buf.ToString();
    }
}
